scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala

```diff
@@ -334,11 +334,218 @@ object OptimizationRules {
           rtp match {
-            case ReduceStepPipe(hg: HashJoinable[K, V2]) =>
+            case ReduceStepPipe(hg: HashJoinable[K @unchecked, V2 @unchecked]) =>
               HashCoGroup(ltp, hg, joiner)
             case otherwise =>
               HashCoGroup(ltp, IdentityReduce(ordK, otherwise, None, Nil), joiner)
           }
         })
     }
+
+
+  /////////////////////////////
+  //
+  // Here are some actual rules for simplifying TypedPipes
+  //
+  /////////////////////////////
+
+  object ComposeFlatMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case FlatMapped(FlatMapped(in, fn0), fn1) =>
+        FlatMapped(in, FlatMappedFn(fn1).runAfter(FlatMapping.FlatM(fn0)))
+      case FlatMapValues(FlatMapValues(in, fn0), fn1) =>
+        FlatMapValues(in, FlatMappedFn(fn1).runAfter(FlatMapping.FlatM(fn0)))
+    }
+  }
+
+  object ComposeMap extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case Mapped(Mapped(in, fn0), fn1) =>
+        Mapped(in, ComposedMapFn(fn0, fn1))
+      case MapValues(MapValues(in, fn0), fn1) =>
+        MapValues(in, ComposedMapFn(fn0, fn1))
+    }
+  }
+
+  object ComposeFilter extends Rule[TypedPipe] {
+    def apply[T](on: Dag[TypedPipe]) = {
+      // scala can't type check this, so we hold its hand:
+      // case Filter(Filter(in, fn0), fn1) =>
+      //   Some(Filter(in, ComposedFilterFn(fn0, fn1)))
+      case f@Filter(_, _) =>
+        def go[A](f: Filter[A]): Option[TypedPipe[A]] =
+          f.input match {
+            case f1: Filter[a] =>
+              Some(Filter[a](f1.input, ComposedFilterFn(f.fn, f.fn)))
+            case _ => None
+          }
+        go(f)
+      case FilterKeys(FilterKeys(in, fn0), fn1) =>
+        Some(FilterKeys(in, ComposedFilterFn(fn0, fn1)))
+      case _ => None
+    }
+  }
+
+  object ComposeWithOnComplete extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case WithOnComplete(WithOnComplete(pipe, fn0), fn1) =>
+        WithOnComplete(pipe, ComposedOnComplete(fn0, fn1))
+    }
+  }
+
+  object RemoveDuplicateForceFork extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case ForceToDisk(ForceToDisk(t)) => ForceToDisk(t)
+      case ForceToDisk(Fork(t)) => ForceToDisk(t)
+      case Fork(Fork(t)) => Fork(t)
+      case Fork(ForceToDisk(t)) => ForceToDisk(t)
+    }
+  }
+
+  /**
+   * We ignore .group if there are is no setting of reducers
+   *
+   * This is arguably not a great idea, but scalding has always
+   * done it to minimize accidental map-reduce steps
+   */
+  object IgnoreNoOpGroup extends PartialRule[TypedPipe] {
+    def applyWhere[T](on: Dag[TypedPipe]) = {
+      case ReduceStepPipe(IdentityReduce(_, input, None, _)) =>
+        input
+    }
+  }
+
+  /**
+   * In map-reduce settings, Merge is almost free in two contexts:
+   * 1. the final write
```

Comment:
This rule i'm not sure in a MR context what the gain is. I think deferring the merge is only useful for filters maybe? since if you have two steps

M -> R1
M -> R2
and do some (R1 ++ R2).filter(...) we know it should be good. But doing flatmaps and maps pushed in there isn't as clear of a win to me?

Reply:
as I mentioned in the comment, if you can defer the merge until a groupBy, scalding can skip doing any work. In tez, if not, that is actually an extra materialization. On MR, cascading applies this same optimization under the hood (but not, for some reason, with Tez).

Also, clearly, for writes, it is a win.

Note, each tuple is only one one side or the other, by pushing it lower (closer to sinks) in the graphs, we don't do more work. So it is hard to see how it hurts also.

Reply:
Which side do the functions goto in a group by, this might change how things get planned on MR negatively?

I.e.
```
M -> R1
M -> R2
(R1 ++ R2).flatMap { huge data cubing function}.sumByKey
```
If the flat map is pushed through the merge its more likely it'll end up on the reducers and you'll write a lot more data. This isn't entirely under the control of scalding in general ofc, but this sort of rule could change how the DAG is exposed to cascading and cause odd regressions?

Ah so the notion here is that if you merge 2 sources say from disk, then do some map operations and a group by. Under tez cascading will introduce an extra step to do the merge pre-map operations? 

For writes its a win under normal MR? is it? 

given only 1 set of mappers or reducers will write to any sink won't the effects be sort of random depending on data sizes? 
```
M -> R(f1,f2) ------> MergeMapperToSink
M -> R2(f1,f2) ---/

M -> R ------> MergeMapperToSink(f1,f2)
M -> R2 ---/
```
Which one of these is better largely depends on the shape of f1 and f2 no? (i.e. input + output data types and how well they serialize/compress, if they expand the data or similar?)

Some examples would be something like writing ngrams from multiple sources, in this case you'd want to merge to be late since you actually write more tuples possibly than you had in either of the previous steps. (Not so common granted, but I did it recently so comes to mind).

Reply:
@ianoc - yeah I think I ran into the merge + map + groupBy issue in cascading 3 - https://github.com/cwensel/cascading/pull/59. From what I remember there was an extra merge node.